### PR TITLE
script: add option for debug build

### PR DESCRIPTION
**Pull request by batrick**
See: https://github.com/ceph/ceph-build/pull/2167

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/script/ptl-tool.py [hunk 1]
```diff
@@ -389,6 +391,7 @@ def main():
     else:
         argv = sys.argv[1:]
     parser.add_argument('--branch', dest='branch', action='store', default=default_branch, help='branch to create ("HEAD" leaves HEAD detached; i.e. no branch is made)')
+    parser.add_argument('--debug-build', dest='debug_build', action='store_true', help='append -debug to branch name prompt ceph-build to build with CMAKE_BUILD_TYPE=Debug')
```
Comment (idryomov) [inline src/script/ptl-tool.py:394]:
```suggestion
    parser.add_argument('--debug-build', dest='debug_build', action='store_true', help='append -debug to branch name prompting ceph-build to build with CMAKE_BUILD_TYPE=Debug')
```

Comment (batrick) [reply src/script/ptl-tool.py:394]: thanks, fixed

**Comment by batrick** [comment]
```
+ [[ wip-batrick-testing-20231006.014828-debug == *-debug ]]
+ CEPH_EXTRA_CMAKE_ARGS+=' -DCMAKE_BUILD_TYPE=Debug -DWITH_CEPH_DEBUG_MUTEX=ON'
+ printf 'Added debug cmake configs to branch %s. CEPH_EXTRA_CMAKE_ARGS: %s\n' wip-batrick-testing-20231006.014828-debug ' -DCMAKE_BUILD_TYPE=Debug -DWITH_CEPH_DEBUG_MUTEX=ON'
Added debug cmake configs to branch wip-batrick-testing-20231006.014828-debug. CEPH_EXTRA_CMAKE_ARGS:  -DCMAKE_BUILD_TYPE=Debug -DWITH_CEPH_DEBUG_MUTEX=ON
```

https://jenkins.ceph.com/job/ceph-dev-new-setup/74098/console

https://shaman.ceph.com/builds/ceph/wip-batrick-testing-20231006.014828-debug/cfbdc475a5ca4098c0330e42cd978c9fd647e012/

centos 8 fails with numerous compiler warnings:

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/73043//consoleFull

**Comment by batrick** [comment]
This PR is ready to go though.

**Comment by idryomov** [comment]
> centos 8 fails with numerous compiler warnings:
> 
> https://shaman.ceph.com/builds/ceph/wip-batrick-testing-20231006.014828-debug/cfbdc475a5ca4098c0330e42cd978c9fd647e012/default/360990/

There is a ton of `-Wdeprecated` noise, but the actual error is in the arrow submodule:
```
[  4%] Building CXX object src/arrow/CMakeFiles/arrow_objlib.dir/array/array_binary.cc.o
In file included from [01m[K/opt/rh/gcc-toolset-11/root/usr/include/c++/11/x86_64-redhat-linux/bits/os_defines.h:39[m[K,
                 from [01m[K/opt/rh/gcc-toolset-11/root/usr/include/c++/11/x86_64-redhat-linux/bits/c++config.h:2732[m[K,
                 from [01m[K/opt/rh/gcc-toolset-11/root/usr/include/c++/11/cstdint:38[m[K,
                 from [01m[K/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-6631-gcfbdc475/rpm/el8/BUILD/ceph-18.0.0-6631-gcfbdc475/src/arrow/cpp/src/arrow/array/array_binary.h:23[m[K,
                 from [01m[K/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/18.0.0-6631-gcfbdc475/rpm/el8/BUILD/ceph-18.0.0-6631-gcfbdc475/src/arrow/cpp/src/arrow/array/array_binary.cc:18[m[K:
[01m[K/usr/include/features.h:381:4:[m[K [01;31m[Kerror: [m[K#warning _FORTIFY_SOURCE requires compiling with optimization (-O) [[01;31m[K-Werror=cpp[m[K]
  381 | #  [01;31m[Kwarning[m[K _FORTIFY_SOURCE requires compiling with optimization (-O)
      |    [01;31m[K^~~~~~~[m[K
cc1plus: all warnings being treated as errors
make[5]: *** [src/arrow/CMakeFiles/arrow_objlib.dir/build.make:90: src/arrow/CMakeFiles/arrow_objlib.dir/array/array_binary.cc.o] Error 1
make[5]: *** Waiting for unfinished jobs....
```
cc @cbodley

**Review by idryomov — APPROVED** [review]
> This PR is ready to go though.

Seems obviously correct to me! (I don't use this script)

**Comment by idryomov** [comment]
jenkins test api

**Comment by cbodley** [comment]
> _FORTIFY_SOURCE requires compiling with optimization (-O)

our build is injecting this `-D_FORTIFY_SOURCE=2` define, so we could remove that from these debug builds. another option is to always build arrow in release mode. in https://github.com/ceph/ceph/blob/080768f/cmake/modules/BuildArrow.cmake#L84-L88, our CMAKE_BUILD_TYPE gets forwarded to the arrow build, but we could leave that on Release instead

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
> > _FORTIFY_SOURCE requires compiling with optimization (-O)
> 
> our build is injecting this `-D_FORTIFY_SOURCE=2` define, so we could remove that from these debug builds. another option is to always build arrow in release mode. in https://github.com/ceph/ceph/blob/080768f/cmake/modules/BuildArrow.cmake#L84-L88, our CMAKE_BUILD_TYPE gets forwarded to the arrow build, but we could leave that on Release instead

https://tracker.ceph.com/issues/63130

**Pull request merged**